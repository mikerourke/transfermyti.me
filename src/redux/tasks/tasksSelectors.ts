import { isNil, propOr } from "ramda";

import { selectIdToLinkedId } from "~/api/selectIdToLinkedId";
import { sourceProjectsByIdSelector } from "~/redux/projects/projectsSelectors";
import { createSelector } from "~/redux/reduxTools";
import { sourceTimeEntryCountByIdFieldSelectorFactory } from "~/redux/timeEntries/timeEntriesSelectors";
import { activeWorkspaceIdSelector } from "~/redux/workspaces/workspacesSelectors";
import type { ReduxState, Task, TaskTableRecord } from "~/typeDefs";

export const sourceTasksByIdSelector = createSelector(
  (state: ReduxState) => state.tasks.source,
  (sourceTasksById): Dictionary<Task> => sourceTasksById,
);

const targetTasksByIdSelector = createSelector(
  (state: ReduxState) => state.tasks.target,
  (targetTasksById): Dictionary<Task> => targetTasksById,
);

export const sourceTasksSelector = createSelector(
  sourceTasksByIdSelector,
  (sourceTasksById): Task[] => Object.values(sourceTasksById),
);

export const includedSourceTasksSelector = createSelector(
  sourceTasksSelector,
  (sourceTasks): Task[] =>
    sourceTasks.filter((sourceTask) => sourceTask.isIncluded),
);

export const includedSourceTasksCountSelector = createSelector(
  includedSourceTasksSelector,
  (includedSourceTasks) => includedSourceTasks.length,
);

export const sourceTasksForTransferSelector = createSelector(
  includedSourceTasksSelector,
  (sourceTasks) =>
    sourceTasks.filter((sourceTask) => isNil(sourceTask.linkedId)),
);

export const sourceTasksInActiveWorkspaceSelector = createSelector(
  activeWorkspaceIdSelector,
  sourceTasksSelector,
  (workspaceId, sourceTasks): Task[] =>
    sourceTasks.filter((task) => task.workspaceId === workspaceId),
);

export const taskIdToLinkedIdSelector = createSelector(
  sourceTasksByIdSelector,
  (sourceTasksById): Dictionary<string> => selectIdToLinkedId(sourceTasksById),
);

export const tasksForInclusionsTableSelector = createSelector(
  (state: ReduxState) => state.allEntities.areExistsInTargetShown,
  sourceTasksInActiveWorkspaceSelector,
  targetTasksByIdSelector,
  sourceProjectsByIdSelector,
  sourceTimeEntryCountByIdFieldSelectorFactory("taskId"),
  (
    areExistsInTargetShown,
    sourceTasks,
    targetTasksById,
    sourceProjectsById,
    timeEntryCountByTaskId,
  ): TaskTableRecord[] => {
    const taskTableRecords: TaskTableRecord[] = [];

    for (const sourceTask of sourceTasks) {
      const existsInTarget = sourceTask.linkedId !== null;
      if (existsInTarget && !areExistsInTargetShown) {
        continue;
      }

      let isActiveInTarget = false;
      if (existsInTarget) {
        const targetId = sourceTask.linkedId as string;

        isActiveInTarget = targetTasksById[targetId].isActive;
      }

      const entryCount = propOr<number, Dictionary<number>, number>(
        0,
        sourceTask.id,
        timeEntryCountByTaskId,
      );

      taskTableRecords.push({
        ...sourceTask,
        entryCount,
        existsInTarget,
        isActiveInSource: sourceTask.isActive,
        isActiveInTarget,
        projectName: sourceProjectsById[sourceTask.projectId].name,
      });
    }

    return taskTableRecords;
  },
);

export const tasksTotalCountsByTypeSelector = createSelector(
  tasksForInclusionsTableSelector,
  (tasksForInclusionsTable) =>
    tasksForInclusionsTable.reduce(
      (
        acc,
        {
          entryCount,
          existsInTarget,
          isIncluded,
          isActiveInSource,
          isActiveInTarget,
        }: TaskTableRecord,
      ) => ({
        entryCount: acc.entryCount + entryCount,
        existsInTarget: acc.existsInTarget + (existsInTarget ? 1 : 0),
        isActiveInSource: acc.isActiveInSource + (isActiveInSource ? 1 : 0),
        isActiveInTarget: acc.isActiveInTarget + (isActiveInTarget ? 1 : 0),
        isIncluded: acc.isIncluded + (isIncluded ? 1 : 0),
      }),
      {
        entryCount: 0,
        existsInTarget: 0,
        isActiveInSource: 0,
        isActiveInTarget: 0,
        isIncluded: 0,
      },
    ),
);
