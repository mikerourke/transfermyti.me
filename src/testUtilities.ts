export { FAKES } from "~/redux/__fakes__/state";

type MockInstance = jest.MockInstance<AnyValid, AnyValid>;

interface WrapForMockingReturn {
  calls: AnyValid[];
  clear(): MockInstance;
  implementOnce(funcOrReturn: AnyValid): MockInstance;
  rejectAlways(err?: AnyValid): MockInstance;
  rejectOnce(err?: AnyValid): MockInstance;
  resolveMultiple(values: AnyValid[]): void;
  resolveOnce(value?: AnyValid): MockInstance;
  returnOnce(value?: AnyValid): MockInstance;
  returnMultiple(values: AnyValid[]): void;
  throwOnce(message?: string): MockInstance;
}

/**
 * Wrappers around someFunc.mock...() that Jest provides. Using
 * this avoids the annoying `someFunc as AnyValid` required to shut TypeScript
 * up.
 */
export function wrapForMocking(func: AnyValid): WrapForMockingReturn {
  return {
    calls: func?.mock?.calls,
    clear(): MockInstance {
      return func.mockClear();
    },
    implementOnce(funcOrReturn: AnyValid): MockInstance {
      if (typeof funcOrReturn === "function") {
        return func.mockImplementationOnce(funcOrReturn);
      } else {
        return func.mockImplementationOnce(() => funcOrReturn);
      }
    },
    rejectAlways(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValue(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    rejectOnce(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValueOnce(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    resolveMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockResolvedValueOnce(value);
      }
    },
    resolveOnce(value: AnyValid = null): MockInstance {
      return func.mockResolvedValueOnce(value);
    },
    returnOnce(value: AnyValid = null): MockInstance {
      return func.mockReturnValueOnce(value);
    },
    returnMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockReturnValueOnce(value);
      }
    },
    throwOnce(message: string = "error"): MockInstance {
      return func.mockImplementationOnce(() => {
        throw new Error(message);
      });
    },
  };
}

/**
 * Allows you to get around `delay()` calls in sagas.
 * @see https://github.com/jfairbank/redux-saga-test-plan/issues/257#issuecomment-503016758
 * @example
 *   return expectSaga(someSaga)
 *     .provide([
 *       { call: provideDelay },
 *     ])
 *     ...
 */
export function provideDelay(
  { fn }: AnyValid,
  next: AnyValid,
): null | VoidFunction {
  return fn.name === "delayP" ? null : next();
}
