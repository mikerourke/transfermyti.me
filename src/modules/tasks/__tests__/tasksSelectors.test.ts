import cases from "jest-in-case";

import { FAKES } from "~/testUtilities";

import * as tasksSelectors from "../tasksSelectors";

const MOCK_STATE = {
  ...FAKES.REDUX_STATE,
  tasks: {
    ...FAKES.REDUX_STATE.tasks,
    source: {
      ...FAKES.REDUX_STATE.tasks.source,
      "7001": {
        ...FAKES.REDUX_STATE.tasks.source["7001"],
        entryCount: 2,
        linkedId: "clock-task-01",
        isIncluded: false,
      },
      "7002": {
        ...FAKES.REDUX_STATE.tasks.source["7002"],
        entryCount: 1,
        linkedId: "clock-task-02",
        isIncluded: false,
      },
      "7003": {
        ...FAKES.REDUX_STATE.tasks.source["7003"],
        entryCount: 1,
        linkedId: null,
        isIncluded: true,
      },
      "7004": {
        ...FAKES.REDUX_STATE.tasks.source["7004"],
        entryCount: 1,
        linkedId: null,
        isIncluded: true,
        isActive: false,
      },
      "7005": {
        ...FAKES.REDUX_STATE.tasks.source["7005"],
        entryCount: 3,
        linkedId: null,
        isIncluded: true,
      },
    },
  },
};

describe("within tasksSelectors", () => {
  test("the sourceTasksByIdSelector returns state.source", () => {
    const result = tasksSelectors.sourceTasksByIdSelector(MOCK_STATE);

    expect(result).toEqual(MOCK_STATE.tasks.source);
  });

  cases(
    "the selectors match their snapshots",
    (options) => {
      const result = options.selector(MOCK_STATE);

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "for the includedSourceTasksSelector",
        selector: tasksSelectors.includedSourceTasksSelector,
      },
      {
        name: "for the sourceTasksForTransferSelector",
        selector: tasksSelectors.sourceTasksForTransferSelector,
      },
      {
        name: "for the sourceTasksInActiveWorkspaceSelector",
        selector: tasksSelectors.sourceTasksInActiveWorkspaceSelector,
      },
      {
        name: "for the taskIdToLinkedIdSelector",
        selector: tasksSelectors.taskIdToLinkedIdSelector,
      },
      {
        name: "for the tasksTotalCountsByTypeSelector",
        selector: tasksSelectors.tasksTotalCountsByTypeSelector,
      },
    ],
  );

  test("the includedSourceTasksCountSelector returns the count of included source tasks", () => {
    const result = tasksSelectors.includedSourceTasksCountSelector(MOCK_STATE);

    expect(result).toBe(3);
  });

  cases(
    "the tasksForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown",
    (options) => {
      const updatedState = {
        ...MOCK_STATE,
        allEntities: {
          ...MOCK_STATE.allEntities,
          areExistsInTargetShown: options.areExistsInTargetShown,
        },
      };

      const result = tasksSelectors.tasksForInclusionsTableSelector(updatedState);

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "when state.allEntities.areExistsInTargetShown = true",
        areExistsInTargetShown: true,
      },
      {
        name: "when state.allEntities.areExistsInTargetShown = false",
        areExistsInTargetShown: false,
      },
    ],
  );
});
