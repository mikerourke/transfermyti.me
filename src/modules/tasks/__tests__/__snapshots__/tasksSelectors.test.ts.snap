// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`within tasksSelectors the selectors match their snapshots for the includedSourceTasksSelector 1`] = `
[
  {
    "assigneeIds": [],
    "entryCount": 1,
    "estimate": "PT6H",
    "id": "7003",
    "isActive": true,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tasks",
    "name": "Task 3",
    "projectId": "2002",
    "workspaceId": "1001",
  },
  {
    "assigneeIds": [],
    "entryCount": 1,
    "estimate": "PT6H",
    "id": "7004",
    "isActive": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tasks",
    "name": "Task 4",
    "projectId": "2003",
    "workspaceId": "1001",
  },
  {
    "assigneeIds": [],
    "entryCount": 3,
    "estimate": "PT6H",
    "id": "7005",
    "isActive": true,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tasks",
    "name": "Task 5",
    "projectId": "2004",
    "workspaceId": "1001",
  },
]
`;

exports[`within tasksSelectors the selectors match their snapshots for the sourceTasksForTransferSelector 1`] = `
[
  {
    "assigneeIds": [],
    "entryCount": 1,
    "estimate": "PT6H",
    "id": "7003",
    "isActive": true,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tasks",
    "name": "Task 3",
    "projectId": "2002",
    "workspaceId": "1001",
  },
  {
    "assigneeIds": [],
    "entryCount": 1,
    "estimate": "PT6H",
    "id": "7004",
    "isActive": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tasks",
    "name": "Task 4",
    "projectId": "2003",
    "workspaceId": "1001",
  },
  {
    "assigneeIds": [],
    "entryCount": 3,
    "estimate": "PT6H",
    "id": "7005",
    "isActive": true,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tasks",
    "name": "Task 5",
    "projectId": "2004",
    "workspaceId": "1001",
  },
]
`;

exports[`within tasksSelectors the selectors match their snapshots for the sourceTasksInActiveWorkspaceSelector 1`] = `
[
  {
    "assigneeIds": [
      "123123",
    ],
    "entryCount": 2,
    "estimate": "PT2H",
    "id": "7001",
    "isActive": true,
    "isIncluded": false,
    "linkedId": "clock-task-01",
    "memberOf": "tasks",
    "name": "Task 1",
    "projectId": "2001",
    "workspaceId": "1001",
  },
  {
    "assigneeIds": [],
    "entryCount": 1,
    "estimate": "PT0M",
    "id": "7002",
    "isActive": true,
    "isIncluded": false,
    "linkedId": "clock-task-02",
    "memberOf": "tasks",
    "name": "Task 2",
    "projectId": "2001",
    "workspaceId": "1001",
  },
  {
    "assigneeIds": [],
    "entryCount": 1,
    "estimate": "PT6H",
    "id": "7003",
    "isActive": true,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tasks",
    "name": "Task 3",
    "projectId": "2002",
    "workspaceId": "1001",
  },
  {
    "assigneeIds": [],
    "entryCount": 1,
    "estimate": "PT6H",
    "id": "7004",
    "isActive": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tasks",
    "name": "Task 4",
    "projectId": "2003",
    "workspaceId": "1001",
  },
  {
    "assigneeIds": [],
    "entryCount": 3,
    "estimate": "PT6H",
    "id": "7005",
    "isActive": true,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tasks",
    "name": "Task 5",
    "projectId": "2004",
    "workspaceId": "1001",
  },
]
`;

exports[`within tasksSelectors the selectors match their snapshots for the taskIdToLinkedIdSelector 1`] = `
{
  "7001": "clock-task-01",
  "7002": "clock-task-02",
  "clock-task-01": "7001",
  "clock-task-02": "7002",
}
`;

exports[`within tasksSelectors the selectors match their snapshots for the tasksTotalCountsByTypeSelector 1`] = `
{
  "entryCount": 8,
  "existsInTarget": 2,
  "isActiveInSource": 4,
  "isActiveInTarget": 1,
  "isIncluded": 3,
}
`;

exports[`within tasksSelectors the tasksForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown when state.allEntities.areExistsInTargetShown = false 1`] = `
[
  {
    "assigneeIds": [],
    "entryCount": 1,
    "estimate": "PT6H",
    "existsInTarget": false,
    "id": "7003",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tasks",
    "name": "Task 3",
    "projectId": "2002",
    "projectName": "Test Project B",
    "workspaceId": "1001",
  },
  {
    "assigneeIds": [],
    "entryCount": 1,
    "estimate": "PT6H",
    "existsInTarget": false,
    "id": "7004",
    "isActive": false,
    "isActiveInSource": false,
    "isActiveInTarget": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tasks",
    "name": "Task 4",
    "projectId": "2003",
    "projectName": "Test Project C",
    "workspaceId": "1001",
  },
  {
    "assigneeIds": [],
    "entryCount": 3,
    "estimate": "PT6H",
    "existsInTarget": false,
    "id": "7005",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tasks",
    "name": "Task 5",
    "projectId": "2004",
    "projectName": "Test Project D",
    "workspaceId": "1001",
  },
]
`;

exports[`within tasksSelectors the tasksForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown when state.allEntities.areExistsInTargetShown = true 1`] = `
[
  {
    "assigneeIds": [
      "123123",
    ],
    "entryCount": 2,
    "estimate": "PT2H",
    "existsInTarget": true,
    "id": "7001",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": true,
    "isIncluded": false,
    "linkedId": "clock-task-01",
    "memberOf": "tasks",
    "name": "Task 1",
    "projectId": "2001",
    "projectName": "Test Project A",
    "workspaceId": "1001",
  },
  {
    "assigneeIds": [],
    "entryCount": 1,
    "estimate": "PT0M",
    "existsInTarget": true,
    "id": "7002",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": false,
    "linkedId": "clock-task-02",
    "memberOf": "tasks",
    "name": "Task 2",
    "projectId": "2001",
    "projectName": "Test Project A",
    "workspaceId": "1001",
  },
  {
    "assigneeIds": [],
    "entryCount": 1,
    "estimate": "PT6H",
    "existsInTarget": false,
    "id": "7003",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tasks",
    "name": "Task 3",
    "projectId": "2002",
    "projectName": "Test Project B",
    "workspaceId": "1001",
  },
  {
    "assigneeIds": [],
    "entryCount": 1,
    "estimate": "PT6H",
    "existsInTarget": false,
    "id": "7004",
    "isActive": false,
    "isActiveInSource": false,
    "isActiveInTarget": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tasks",
    "name": "Task 4",
    "projectId": "2003",
    "projectName": "Test Project C",
    "workspaceId": "1001",
  },
  {
    "assigneeIds": [],
    "entryCount": 3,
    "estimate": "PT6H",
    "existsInTarget": false,
    "id": "7005",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tasks",
    "name": "Task 5",
    "projectId": "2004",
    "projectName": "Test Project D",
    "workspaceId": "1001",
  },
]
`;
